### PR TITLE
NFT Database Schema

### LINKED ISSUE #2056: NFT database schema
**Problem**
We need nft database added to unblock other tasks.

**Solution**
Add schema as outlined in [design doc](https://github.com/hashgraph/hedera-mirror-node/blob/master/docs/design/nft.md).

**Alternatives**

**Additional Context**

### PR DESCRIPTION
**Pull request by ijungmann**
**Detailed description**:
Add Flyway migrations to support NFTs (based on NFT design doc)

- Update `token` table with new values
- Add `nft` and `nft_transfer`  tables, indexes, etc.
- Add new Response Codes

**Which issue(s) this PR fixes**:
Fixes #2056 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [ ] Tests updated

### CODE REVIEW

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql [hunk 1]
```diff
@@ -0,0 +1,44 @@
+-------------------
+-- Update and add tables for NFT support
+-------------------
+
+-- Update the token table
+alter table token
+    add column max_supply bigint not null default 9223372036854775807,
```
Comment (ijungmann) [inline hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql:7]: Max long

File: hedera-mirror-importer/src/main/resources/db/migration/v2/V2.0.0__time_scale_init.sql [hunk 2]
```diff
@@ -310,3 +337,4 @@ create table if not exists transaction
     transaction_bytes      bytea
 );
 comment on table transaction is 'Submitted network transactions';
+
```
Comment (ijungmann) [inline hedera-mirror-importer/src/main/resources/db/migration/v2/V2.0.0__time_scale_init.sql:340]: Remove

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  serial_number         bigint  not null,
+  token_id              bigint  not null
+);
+create unique index if not exists nft_transfer__timestamp_token_id_serial_num
+    on nft_transfer (consensus_timestamp, token_id, serial_number);
+comment on table nft_transfer is 'Crypto account nft transfers';
+
+-- Insert new response codes
+insert into t_transaction_results (result, proto_id)
+values ('ACCOUNT_EXPIRED_AND_PENDING_REMOVAL', 223),
```
Comment (ijungmann) [inline hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql:41]: This one isn't part of the NFT protobufs change, but I noticed we were missing it.

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql [hunk 4]
```diff
@@ -0,0 +1,44 @@
+-------------------
+-- Update and add tables for NFT support
+-------------------
+
+-- Update the token table
+alter table token
+    add column max_supply bigint not null default 9223372036854775807, -- max long
+    add column supply_type character varying(10) not null default 'INFINITE',
+    add column type character varying(20) not null default 'NON_FUNGIBLE_UNIQUE';
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql:9]: I believe the default should be `FUNGIBLE_COMMON`
```suggestion
    add column type character varying(20) not null default 'FUNGIBLE_COMMON';
```

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  created_timestamp     bigint  primary key     not null,
+  deleted               boolean default false   not null,
+  modified_timestamp    bigint                  not null,
+  metadata              bytea   default ''      not null,
+  serial_number         bigint                  not null,
+  token_id              bigint                  not null
+);
+create unique index if not exists nft__token_id_serial_num
+    on nft (token_id, serial_number);
+comment on table nft is 'NFT';
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql:24]: nit: Something more suggestive. Template suggestions below that you can modify.
```suggestion
comment on table nft is 'Non Fungible Tokens (NFTs) minted on network';
```

File: hedera-mirror-importer/src/main/resources/db/migration/v2/V2.0.0__time_scale_init.sql [hunk 6]
```diff
@@ -228,13 +252,16 @@ create table if not exists token
...
+    max_supply             bigint                 not null default 9223372036854775807, -- max long
     modified_timestamp     bigint                 not null,
     name                   character varying(100) not null,
     supply_key             bytea,
     supply_key_ed25519_hex varchar                null,
+    supply_type            character varying(10)  not null default 'INFINITE',
     symbol                 character varying(100) not null,
     total_supply           bigint                 not null default 0,
     treasury_account_id    bigint                 not null,
+    type                   character varying(20) not null default 'NON_FUNGIBLE_UNIQUE',
```
Comment (Nana-EC) [inline hedera-mirror-importer/src/main/resources/db/migration/v2/V2.0.0__time_scale_init.sql:264]: Same thing above, this should be `FUNGIBLE_COMMON` right?
```suggestion
    type                   character varying(20) not null default 'FUNGIBLE_COMMON',
```

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Some discrepancies on the `type`. Also left some nits.

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+create table if not exists nft_transfer
+(
+  consensus_timestamp   bigint  not null,
+  receiver_account_id   bigint  not null,
+  sender_account_id     bigint  not null,
+  serial_number         bigint  not null,
+  token_id              bigint  not null
+);
+create unique index if not exists nft_transfer__timestamp_token_id_serial_num
+    on nft_transfer (consensus_timestamp, token_id, serial_number);
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql:36]: Should be descending. Same for v2

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql:9]: I vote smallint type to be conformant with other enum types we have right now.
If we ever go enumerated we should do it for all.

[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql:9]: I disagree. If we wait for perfection we'll never get anything done. Better to improve things while we can.

Comment (ijungmann) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql]: Correct, not sure why I got that swapped.  Fixed.

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql (hunk 5 above)
Comment (ijungmann) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql]: I like that, just made `Non Fungible` into `Non-Fungible`

File: hedera-mirror-importer/src/main/resources/db/migration/v2/V2.0.0__time_scale_init.sql (hunk 6 above)
Comment (ijungmann) [reply hedera-mirror-importer/src/main/resources/db/migration/v2/V2.0.0__time_scale_init.sql]: Yep, fixed.

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (ijungmann) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql]: Sure, I agree, I like the enumarated type.  Added those, looks like it's working

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql [hunk 8]
```diff
@@ -0,0 +1,44 @@
+-------------------
+-- Update and add tables for NFT support
+-------------------
+
+-- Update the token table
+alter table token
+    add column max_supply bigint not null default 9223372036854775807, -- max long
+    add column supply_type character varying(10) not null default 'INFINITE',
```
[the remark this answers is not in the document]
Comment (ijungmann) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql]: Fixed.

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  account_id            bigint                  not null,
+  created_timestamp     bigint  primary key     not null,
+  deleted               boolean default false   not null,
+  modified_timestamp    bigint                  not null,
+  metadata              bytea   default ''      not null,
+  serial_number         bigint                  not null,
+  token_id              bigint                  not null
+);
+create unique index if not exists nft__token_id_serial_num
+    on nft (token_id, serial_number);
```
[the remark this answers is not in the document]
Comment (ijungmann) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql]: Added

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql (hunk 7 above)
Comment (ijungmann) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql]: Added

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql]: Sure, it's a note for consistency. 
I'm open to enumerated if possible. 
If we do then let's open a ticket to transform the rest over in a future sprint or do it in v2 of the schema.

File: hedera-mirror-importer/src/main/resources/db/migration/v2/V2.0.0__time_scale_init.sql [hunk 10]
```diff
@@ -3,6 +3,10 @@
 -- Supports mirror nodes migrated from v1.0
 -------------------
 
+-- Create enums for tables
+CREATE TYPE token_supply_type AS ENUM ('INFINITE', 'FINITE');
+CREATE TYPE token_type AS ENUM ('FUNGIBLE_COMMON', 'NON_FUNGIBLE_UNIQUE');
+
```
Comment (ijungmann) [inline hedera-mirror-importer/src/main/resources/db/migration/v2/V2.0.0__time_scale_init.sql:9]: Any opinions on creating all enums at the top vs. creating them in their corresponding table sections?

File: hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (ijungmann) [reply hedera-mirror-importer/src/main/resources/db/migration/v1/V1.39.0__support_nfts.sql]: Sure, I can create a ticket for that.

**Pull request merged**